packages/airnode-protocol/contracts/rrp/authorizers/admin/AirnodeTokenPayment.sol

```diff
@@ -0,0 +1,286 @@
+        require(
+            requesterAuthorizerWithManager != address(0),
+            "No RequesterAuthorizersWithManager set for chainId"
+        );
+
+        (
+            uint64 expirationTimestamp,
+            uint192 indefiniteWhitelistCount
+        ) = IRequesterAuthorizerWithManager(requesterAuthorizerWithManager)
+                .airnodeToEndpointIdToRequesterToWhitelistStatus(
+                    _airnode,
+                    _endpointId,
+                    _requesterAddress
+                );
+        require(
+            indefiniteWhitelistCount == 0,
+            "Already whitelisted indefinitely"
+        );
+        require(
+            uint64(block.timestamp) + _whitelistDuration > expirationTimestamp,
+            "Already whitelisted"
+        );
+
+        uint256 amount = getPaymentAmount(
+            _chainId,
+            _airnode,
+            _endpointId,
+            expirationTimestamp > uint64(block.timestamp)
+                ? (uint64(block.timestamp) + _whitelistDuration) -
+                    expirationTimestamp
+                : _whitelistDuration
+        );
+
+        IRequesterAuthorizerWithManager(requesterAuthorizerWithManager)
+            .extendWhitelistExpiration(
+                _airnode,
+                _endpointId,
+                _requesterAddress,
+                uint64(block.timestamp) + _whitelistDuration
+            );
+
+        assert(
+            IERC20Metadata(paymentTokenAddress).transferFrom(
+                msg.sender,
+                getAirnodeToPaymentDestination(_airnode),
+                amount
+            )
+        );
+
+        emit MadePayment(
+            _chainId,
+            _airnode,
+            _endpointId,
+            _requesterAddress,
+            msg.sender,
+            getAirnodeToPaymentDestination(_airnode),
+            amount,
+            IERC20Metadata(paymentTokenAddress).symbol(),
+            uint64(block.timestamp) + _whitelistDuration
+        );
+    }
+
+    /// @notice Returns the amount of tokens a sponsor has to transfer to the
+    /// Airnode in order to be whitelisted for a given
+    /// chainId-airnode-endpointId
+    /// @dev AirnodeFeeRegistry should return the fee in usd using 18 decimals
+    /// @param _chainId Id of the chain
+    /// @param _airnode Airnode address
+    /// @param _endpointId Id of the endpoint
+    /// @param _whitelistDuration Duration in seconds for which the requester
+    /// will be whitelisted
+    function getPaymentAmount(
+        uint256 _chainId,
+        address _airnode,
+        bytes32 _endpointId,
+        uint64 _whitelistDuration
+    ) public view override returns (uint256 amount) {
+        uint256 feeInUsd = IAirnodeFeeRegistry(airnodeFeeRegistry)
+            .getEndpointPrice(_chainId, _airnode, _endpointId);
+        uint24 feeInterval = IAirnodeFeeRegistry(airnodeFeeRegistry).INTERVAL();
+        amount =
+            (feeInUsd * _whitelistDuration) /
```

Comment:
Im thinking about this more and it seems like we have to multiply this with decimal places of erc20 because otherwise the payment amount isn't right
```
feeInUsd = 100 * 10**18
whitelistduration = 100
paymentTokenPrice = 1 * 10**18
feeInterval = 1000

amount = (100 * 10**18) * 100 / ((1 * 10**18) * 1000) = 10 

```

but it should give 10 * (decimal places of erc20)

important to keep the decimal places because different erc20s have different decimal places.

Reply:
True. I'll fix it by fetching the decimals from the payment token like we were doing on the initial draft of this contract.

Reply:
this seem alright, I guess I have just one small caveat with this approach, right now we assume that the fee we receive form the AirnodeFeeRegistry contract is implicitly going to be 18 places. If for some reason we plan to deploy a new contract with different decimal places , it will break this logic so for that reason what we can do is multiply by 1e18 to cancel the paymentToken price decimals and divide by the decimals from the feeRegistry contract .

Reply:
Maybe this is not even needed tbh...

Reply:
Right, but since in the last protocol we'd decided that everything will be expressed in 18 decimals I made that assumption that fee and price would always be in 18 decimal places. Also, now that I think about it I don't even need to use the token decimals in the formula 🤔 I could just have a multiplier coefficient and multiply the numerator by that coefficient in order to make sure I get 18 decimals back from the integer division.

Reply:
I don't think I am completely following but essentially the payment will be made in using the erc20 decimals, so if an erc20 has 6 decimals and we end up using 18 decimal places then we will be specifying the wrong amount by a large magnitude.

Reply:
I think I now get what you mean 👍🏻

Reply:
> Right, but since in the last protocol we'd decided that everything will be expressed in 18 decimals I made that assumption that fee and price would always be in 18 decimal places. Also, now that I think about it I don't even need to use the token decimals in the formula thinking I could just have a multiplier coefficient and multiply the numerator by that coefficient in order to make sure I get 18 decimals back from the integer division.

Implemented it [here](https://github.com/api3dao/airnode/pull/656/commits/2ea56225ee6cbdccbb1ce7ad86c290a27c1c3a6b)